### ISSUE TITLE
DeferredResource does not  return correct isLeaf resulting in bad postpath

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7567]
|<img alt="alexjacobson's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| alexjacobson reported|
|-|-|
|Trac ID|trac#295|
|Type|defect|
|Created|2003-09-29 08:02:12Z|

<details><summary>Searchable metadata</summary>

```
trac-id__295 295
type__defect defect
reporter__alexjacobson alexjacobson
priority__high high
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__ 
keywords__ 
time__1064822532000000 1064822532000000
changetime__1467802175566588 1467802175566588
version__ 
owner__ 
cc__radix cc__alexjacobson
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="alexjacobson's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|alexjacobson commented|
|-|-|

```
#!html
<pre>
Here is a really simple use of DeferredResource
class SDResource(resource.Resource):
    def __init__(self,default):  self.default=default
    def getChildWithDefault(self,name,request):
        d=defer.Deferred()
        reactor.callLater(0,d.callback,self.default)
        return DeferredResource( d
).getChildWithDefault(name, request)

Note that a resource wrapped in this does not get a correct
postpath.

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
I'm confused: what does "return correct isLeaf" mean?

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Please test the following patch, and radix, please write
tests (I can explain the issue to you  if you want).

*** util.py.~1.17.~	Tue Jul 22 13:11:50 2003
--- util.py	Mon Sep 29 20:37:37 2003
***************
*** 94,100 ****
      I wrap up a Deferred that will eventually result in a
Resource
      object.
      """
! 
      def __init__(self, d):
          resource.Resource.__init__(self)
          self.d = d
--- 94,101 ----
      I wrap up a Deferred that will eventually result in a
Resource
      object.
      """
!     isLeaf = 1
!     
      def __init__(self, d):
          resource.Resource.__init__(self)
          self.d = d
***************
*** 106,126 ****
          return NOT_DONE_YET
  
      def _cbChild(self, child, request):
!         request.render(child)
!         return child
  
      def _ebChild(self, reason, request):
          request.processingFailed(reason)
          return reason
  
-     def getChildWithDefault(self, path, request):
-         self.d.addCallback(self._cbGetChild, path, request)
-         return DeferredResource(self.d)
- 
-     def _cbGetChild(self, child, path, request):
-         return child.getChildWithDefault(path, request)
- 
- 
  
  stylesheet = """
  &lt;style type="text/css">
--- 107,124 ----
          return NOT_DONE_YET
  
      def _cbChild(self, child, request):
!         result =
child.getChildForRequest(request).render(request)
!         from twisted.web.server import NOT_DONE_YET
!         if result == NOT_DONE_YET:
!             return
!         else:
!             request.write(result)
!             request.finish()
  
      def _ebChild(self, reason, request):
          request.processingFailed(reason)
          return reason
  
  
  stylesheet = """
  &lt;style type="text/css">

</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
A better patch:

*** util.py.~1.17.~	Tue Jul 22 13:11:50 2003
--- util.py	Fri Oct  3 03:23:00 2003
***************
*** 94,104 ****
      I wrap up a Deferred that will eventually result in a Resource
      object.
      """
! 
      def __init__(self, d):
          resource.Resource.__init__(self)
          self.d = d
  
      def render(self, request):
          self.d.addCallback(self._cbChild, request).addErrback(
              self._ebChild,request)
--- 94,109 ----
      I wrap up a Deferred that will eventually result in a Resource
      object.
      """
!     isLeaf = 1
!     
      def __init__(self, d):
          resource.Resource.__init__(self)
          self.d = d
  
+     def getChild(self, name, request):
+         if name == "":
+             return self
+     
      def render(self, request):
          self.d.addCallback(self._cbChild, request).addErrback(
              self._ebChild,request)
***************
*** 106,126 ****
          return NOT_DONE_YET
  
      def _cbChild(self, child, request):
!         request.render(child)
!         return child
  
      def _ebChild(self, reason, request):
          request.processingFailed(reason)
          return reason
  
-     def getChildWithDefault(self, path, request):
-         self.d.addCallback(self._cbGetChild, path, request)
-         return DeferredResource(self.d)
- 
-     def _cbGetChild(self, child, path, request):
-         return child.getChildWithDefault(path, request)
- 
- 
  
  stylesheet = """
  &lt;style type="text/css">
--- 111,128 ----
          return NOT_DONE_YET
  
      def _cbChild(self, child, request):
!         result = child.getChildForRequest(request).render(request)
!         from twisted.web.server import NOT_DONE_YET
!         if result == NOT_DONE_YET:
!             return
!         else:
!             request.write(result)
!             request.finish()
  
      def _ebChild(self, reason, request):
          request.processingFailed(reason)
          return reason
  
  
  stylesheet = """
  &lt;style type="text/css">
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Use unified diff, not context diff, when submitting patches.
Thanks
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Take three, yet another bug fix. I HATE YOU MILKMAN WEB.

Index: twisted/web/util.py
===================================================================
RCS file: /cvs/Twisted/twisted/web/util.py,v
retrieving revision 1.17
diff -u -r1.17 util.py
--- twisted/web/util.py	21 Jul 2003 04:09:52 -0000	1.17
+++ twisted/web/util.py	5 Oct 2003 00:26:29 -0000
@@ -94,11 +94,15 @@
     I wrap up a Deferred that will eventually result in a Resource
     object.
     """
-
+    isLeaf = 1
+    
     def __init__(self, d):
         resource.Resource.__init__(self)
         self.d = d
 
+    def getChild(self, name, request):
+        return self
+    
     def render(self, request):
         self.d.addCallback(self._cbChild, request).addErrback(
             self._ebChild,request)
@@ -106,20 +110,17 @@
         return NOT_DONE_YET
 
     def _cbChild(self, child, request):
-        request.render(child)
-        return child
+        result = child.getChildForRequest(request).render(request)
+        from twisted.web.server import NOT_DONE_YET
+        if result == NOT_DONE_YET:
+            return
+        else:
+            request.write(result)
+            request.finish()
 
     def _ebChild(self, reason, request):
         request.processingFailed(reason)
         return reason
-
-    def getChildWithDefault(self, path, request):
-        self.d.addCallback(self._cbGetChild, path, request)
-        return DeferredResource(self.d)
-
-    def _cbGetChild(self, child, path, request):
-        return child.getChildWithDefault(path, request)
-
 
 
 stylesheet = """
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
Here's a unit test. I haven't checked whether itamar's patch solves this,
though

Index: twisted/test/test_web.py
===================================================================
RCS file: /cvs/Twisted/twisted/test/test_web.py,v
retrieving revision 1.41
diff -u -r1.41 test_web.py
--- twisted/test/test_web.py    4 Oct 2003 18:30:55 -0000       1.41
+++ twisted/test/test_web.py    5 Oct 2003 13:41:33 -0000
@@ -16,7 +16,7 @@
 from twisted.trial import unittest
 import string, random, copy

-from twisted.web import server, resource, widgets, guard
+from twisted.web import server, resource, widgets, guard, util
 from twisted.internet import app, defer
 from twisted.cred import service, identity, perspective, authorizer
 from twisted.protocols import http, loopback
@@ -411,3 +411,19 @@
         request.gotLength(0)
         request.requestReceived('GET', '/foo/bar', 'HTTP/1.0')
         self.assertEqual(request.prePathURL(), 'https://example.com:81/foo/bar')
+
+class SDResource(resource.Resource):
+    def __init__(self,default):  self.default=default
+    def getChildWithDefault(self,name,request):
+        d=defer.succeed(self.default)
+        return util.DeferredResource(d).getChildWithDefault(name, request)
+
+class SDTest(unittest.TestCase):
+
+    def testDeferredResource(self):
+        r = resource.Resource()
+        r.isLeaf = 1
+        s = SDResource(r)
+        d = DummyRequest(['foo', 'bar', 'baz'])
+        s.getChildForRequest(d)
+        self.assertEqual(d.postpath, ['bar', 'baz'])
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Yay moshez, my code did indeed make the test pass. Commited to CVS.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|GitHub commented|
|-|-|

In changeset 5c447e3663c5aa405bddff7da5577ca1fd81794d
```
#!CommitTicketReference repository="" revision="5c447e3663c5aa405bddff7da5577ca1fd81794d"
Merge pull request #295 from twisted/8209-acabhishek942-default

Author: acabhishek942
Reviewer: rodrigc
Fixes: #8209

Fix docstring to reflect correct default interface value
```
